src/Raven.Server/Documents/ShardedTcpHandlers/ShardedSubscriptionConnection.cs

```diff
@@ -0,0 +1,583 @@
+            foreach (var worker in _workers)
+            {
+                using (_serverStore.ContextPool.AllocateOperationContext(out JsonOperationContext context))
+                using (context.GetMemoryBuffer(out var buffer))
+                {
+                    var receivedMessage = await ReadNextObject(context, worker, buffer);
+                    if (receivedMessage == null || CancellationTokenSource.IsCancellationRequested)
+                        return;
+
+                    if (receivedMessage.Type != SubscriptionConnectionServerMessage.MessageType.ConnectionStatus ||
+                        receivedMessage.Status != SubscriptionConnectionServerMessage.ConnectionStatus.Accepted)
+                    {
+                        using (var bjro = DocumentConventions.Default.Serialization.DefaultConverter.ToBlittable(receivedMessage, context))
+                        {
+                            await WriteBlittableAsync(bjro, TcpConnection);
+                        }
+                    }
+                }
+            }
+
+            await WriteJsonAsync(new DynamicJsonValue
+            {
+                [nameof(SubscriptionConnectionServerMessage.Type)] = nameof(SubscriptionConnectionServerMessage.MessageType.ConnectionStatus),
+                [nameof(SubscriptionConnectionServerMessage.Status)] = nameof(SubscriptionConnectionServerMessage.ConnectionStatus.Accepted)
+            }, TcpConnection);
+        }
+
+        private async Task ProcessSubscriptionAsync()
+        {
+            var replyFromClientTask = GetReplyFromClientAsync(TcpConnection, _copiedBuffer.Buffer, _isDisposed);
+            using (DisposeOnDisconnect)
+            {
+
+                while (CancellationTokenSource.IsCancellationRequested == false)
+                {
+                    foreach (SubscriptionServerWorker worker in _workers)
+                    {
+                        if (CancellationTokenSource.IsCancellationRequested)
+                            break;
+
+                        if (worker.HasPullingTask)
+                            continue;
+
+                        worker.PullingTask = ReadNextObject(worker.Context, worker, worker.Buffer);
+                    }
+
+                    var timeoutTask = TimeoutManager.WaitFor(TimeSpan.FromMilliseconds(3000));
+                    var completedTask = await Task.WhenAny(new[] { replyFromClientTask, timeoutTask }.Concat(_workers.Select(x => x.PullingTask).ToList()));
+
+                    if (CancellationTokenSource.IsCancellationRequested)
+                        break;
+
+                    if (completedTask == replyFromClientTask)
+                    {
+                        // forward to shards
+                        var clientReply = await replyFromClientTask;
+                        foreach (var worker in _workers)
+                        {
+                            await SendClientReplyToWorker(worker, clientReply);
+                        }
+
```

Comment:
This is wrong. You are assuming that a client reply can be sent to all the shards, but that isn't the case.